### DISCUSSION TITLE
how to make helix work with c++ and cmake

### DISCUSSION THREAD

**Discussion by ghost** [helix-editor/helix#5044] [Q&A] [answered]
I'm trying to make a c++ app using cmake as the build system but it seems like helix is failing at understanding my include paths for some reason.

<img width="1440" alt="Screenshot 2022-12-07 at 17 29 48" src="https://user-images.githubusercontent.com/81249153/206205650-82861f18-a684-496d-b127-5c8eb42e43a9.png">

Is there a way to get rid of those red dots without including headers with absolute path?

**Comment by the-mikedavis** [comment]
`clangd` looks for a compilation database (`compile_commands.json` or `compile_flags.txt`) that tells it how to compile the project and find includes. See https://clangd.llvm.org/design/compile-commands#where-do-compile-commands-come-from

I'm not very familiar with this myself but you should be able to setup CMake to emit this https://cmake.org/cmake/help/latest/variable/CMAKE_EXPORT_COMPILE_COMMANDS.html, or use [Bear](https://github.com/rizsotto/Bear)

**Comment by sinefabula** [comment, marked as answer]
For small programs I keep a simple `compile_flags.txt` in my $HOME directory which just lists common search paths (I'm on macOS so I need to specify paths to boost etc.) and my usual compile flags. It looks like this:

```
-std=c++17
-Wall
-I/opt/homebrew/opt/boost/include
-I/opt/homebrew/include
-L/opt/homebrew/lib
```

For larger projects, I usually have a simple script which just runs `cmake` with `-DCMAKE_EXPORT_COMPILE_COMMANDS=ON` to generate `compile_commands.json` and then deletes the CMake cache.

**Reply by ghost** [reply]
Thanks you so much after exporting the file from cmake, moving the `compile_commands.json` to the root of my project directory solved the problem.

**Comment by goyalyashpal** [comment]
it's also not detecting my #include `iostream` and namespace `std`, hence, all the `cout` etc are also shown undeclared

**Comment by AlexisTM** [comment]
In my case, I just added the following in my helix `.helix/languages.toml`. 
This ensures the `compile_commands.json` file is always up to date 👍🏼 

```toml
[[language]]
name = "cpp"
language-server = { command = "clangd", args=["--compile-commands-dir=compile_commands_directory"] }
```

**Comment by ZennMystic** [comment]
So I come across this post when I was looking into trying out Helix for C/C++ programming. I have clangd and llbd-vscode working. But what now?

Way back when I used Vim, I have a key bind of alt-m that would run ./build.sh %... % being the C file in question. The build script would have the compiler flags I wanted and the compiler I wanted to use and were I wanted the output file to go with its name all that. But I read Helix does not have % for filename.

How do you do the same type of thing in Helix?

is this were:

[[language]]
name = "cpp"
language-server = { command = "clangd", args=["--compile-commands-dir=compile_commands_directory"] }

comes in?

Right now I am doing ctrl-z running my build script then fg to go back to /helix.

I am confused to say the least. So any help would be good. Thank for your time.

**Reply by AlexisTM** [reply]
When you use cmake, you can generate the compile commands file (a big JSON file). This file stores all compile commands for every single target of your project.

You can use: `-DCMAKE_EXPORT_COMPILE_COMMANDS=ON` in your cmake configure step so on the next build you get a compile commands file. You can then link the build directory containing that file with the configuration above.

https://cmake.org/cmake/help/latest/variable/CMAKE_EXPORT_COMPILE_COMMANDS.html

**Reply by ky3ow** [reply]
Don't know how to make it single keybind but you can do `:sh ./build.sh <C-r>%`
`:sh` is alias for run-shell-command 
`<C-r>%` inserts `%` registry(full file path) to command prompt

**Comment by ristomatti** [comment]
In my case no additional config was needed, the actual problem was this https://stackoverflow.com/questions/74785927/clangd-doesnt-recognize-standard-headers
